src/local_cluster.rs

```diff
@@ -0,0 +1,146 @@
+use crate::blocktree::{create_new_tmp_ledger, tmp_copy_blocktree};
+use crate::client::mk_client;
+use crate::cluster_info::{Node, NodeInfo};
+use crate::fullnode::{Fullnode, FullnodeConfig};
+use crate::gossip_service::discover;
+use crate::thin_client::retry_get_balance;
+use crate::thin_client::ThinClient;
+use crate::voting_keypair::VotingKeypair;
+use solana_sdk::genesis_block::GenesisBlock;
+use solana_sdk::pubkey::Pubkey;
+use solana_sdk::signature::{Keypair, KeypairUtil};
+use solana_sdk::system_transaction::SystemTransaction;
+use std::fs::remove_dir_all;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::thread::JoinHandle;
+
+pub struct LocalCluster {
+    /// Keypair with funding to particpiate in the network
+    pub funding_keypair: Keypair,
+    /// Entry point from which the rest of the network can be discovered
+    pub entry_point_info: NodeInfo,
+    fullnode_hdls: Vec<(JoinHandle<()>, Arc<AtomicBool>)>,
+    ledger_paths: Vec<String>,
+}
+
+impl LocalCluster {
+    pub fn new(num_nodes: usize, cluster_lamports: u64, lamports_per_node: u64) -> Self {
+        let leader_keypair = Arc::new(Keypair::new());
+        let leader_pubkey = leader_keypair.pubkey();
+        let leader_node = Node::new_localhost_with_pubkey(leader_keypair.pubkey());
+        let (genesis_block, mint_keypair) =
+            GenesisBlock::new_with_leader(cluster_lamports, leader_pubkey, lamports_per_node);
+        let (genesis_ledger_path, _last_id) = create_new_tmp_ledger!(&genesis_block);
+        let leader_ledger_path = tmp_copy_blocktree!(&genesis_ledger_path);
+        let mut ledger_paths = vec![];
+        ledger_paths.push(genesis_ledger_path.clone());
+        ledger_paths.push(leader_ledger_path.clone());
+        let voting_keypair = VotingKeypair::new_local(&leader_keypair);
+        let fullnode_config = FullnodeConfig::default();
+        let leader_node_info = leader_node.info.clone();
+        let leader_server = Fullnode::new(
+            leader_node,
+            &leader_keypair,
+            &leader_ledger_path,
+            voting_keypair,
+            None,
+            &fullnode_config,
+        );
+        let (thread, exit, _) = leader_server.start(None);
+        let mut fullnode_hdls = vec![(thread, exit)];
+        let mut client = mk_client(&leader_node_info);
+        for _ in 0..(num_nodes - 1) {
+            let validator_keypair = Arc::new(Keypair::new());
+            let voting_keypair = VotingKeypair::new_local(&validator_keypair);
+            let validator_pubkey = validator_keypair.pubkey();
+            let validator_node = Node::new_localhost_with_pubkey(validator_keypair.pubkey());
+            let ledger_path = tmp_copy_blocktree!(&genesis_ledger_path);
+            ledger_paths.push(ledger_path.clone());
+
+            // Send each validator some tokens to vote
+            let validator_balance = Self::transfer(
```

Comment:
I think it's not creating the vote accounts for the validators. Now the balance of vote account is used to calculate node stakes (for leader rotations etc).

Reply:
@pgarg66 can you show me?

Reply:
This function (or something similar) needs to be called here for every validator node

https://github.com/solana-labs/solana/blob/master/fullnode/src/main.rs#L51

Reply:
This PR changes the way stakes are calculated https://github.com/solana-labs/solana/pull/2992
Earlier it was using node's account balance. Now it's using staking (voting) account balance.

Reply:
@pgarg66 cool! can you add it?

Reply:
👍